src/datumaro/components/prune.py

```diff
@@ -0,0 +1,322 @@
+                        self._labels.append(annotation.label)
+                    if isinstance(annotation, HashKey):
+                        hash_key = annotation.hash_key[0]
+                        if self._hash_type == "txt":
+                            inputs = category_dict.get(item.annotations[0].label)
+                            if isinstance(inputs, List):
+                                inputs = " ".join(inputs)
+                            hash_key_txt = self._text_model.infer_text(inputs)
+                            hash_key = np.concatenate([hash_key, hash_key_txt])
+                        hash_key = np.unpackbits(hash_key, axis=-1)
+                        if self._database_keys is None:
+                            self._database_keys = hash_key.reshape(1, -1)
+                        else:
+                            self._database_keys = np.concatenate(
+                                (self._database_keys, hash_key.reshape(1, -1)), axis=0
+                            )
+                self._item_list.append(item)
+
+    @property
+    def model(self):
+        if self._model is None:
+            self._model = ExplorerLauncher(model_name="clip_visual_ViT-B_32")
+        return self._model
+
+    @property
+    def text_model(self):
+        if self._text_model is None:
+            self._text_model = ExplorerLauncher(model_name="clip_text_ViT-B_32")
+        return self._text_model
+
+    def prompting(self):
+        category_dict = {}
+        detected_format = self._dataset.format
+        template = format_templates.get(detected_format, templates)
+        for label, indice in list(self._dataset.categories().values())[0]._indices.items():
+            category_dict[indice] = [temp.format(label) for temp in template]
+        return category_dict
+
+    def _compute_hash_key(self, datasets, datasets_to_infer):
+        for dataset_to_infer in datasets_to_infer:
+            if dataset_to_infer:
+                dataset_to_infer.run_model(self.model, append_annotation=True)
+        for dataset, dataset_to_infer in zip(datasets, datasets_to_infer):
+            updated_items = [
+                dataset.get(item.id, item.subset).wrap(annotations=item.annotations)
+                for item in dataset_to_infer
+            ]
+            dataset.update(updated_items)
+        return datasets
+
+    def get_pruned(self, ratio: float = 0.5) -> Dataset:
+        method = {
+            "random": random_select,
+            "cluster_random": clustered_random,
+            "centroid": centroid,
+            "query_clust": query_clust,
+            "entropy": entropy,
+            "ndr": ndr_select,
+        }
+
+        selected_items, dist_tuples = method[self._cluster_method](
+            ratio=ratio,
+            num_centers=self._num_centers,
+            labels=self._labels,
+            database_keys=self._database_keys,
+            item_list=self._item_list,
+            source=self._dataset,
+        )
```

Comment:
Please make a common ancestor having abstractmethod for those methods. It also should have correct typing for the input arguments.

Reply:
I made `PruneBase` and override `base` for each method. I could not cover different input arguments for each method. If you have any suggestions on this part, I'd appreciate it if you could let me know.